### PR TITLE
MODEMAIL-70: Logging improvement

### PR DESCRIPTION
**Pull request by MurodjonMukhamadaliev**
https://issues.folio.org/browse/MODEMAIL-70

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java [hunk 1]
```diff
@@ -36,6 +36,7 @@ public DelayedTasksAPI(Vertx vertx, String tenantId) {
   @Override
   public void deleteDelayedTaskExpiredMessages(String expirationDate, String status,
     Map<String, String> headers, Handler<AsyncResult<Response>> resultHandler, Context context) {
+    logger.debug("deleteDelayedTaskExpiredMessages:: parameters: expirationDate={}, status={}", expirationDate, status);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/DelayedTasksAPI.java:39]: - blank line above
- headers should be also logged

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java [hunk 2]
```diff
@@ -50,11 +51,11 @@ public void deleteDelayedTaskExpiredMessages(String expirationDate, String statu
   @Override
   public void postDelayedTaskRetryFailedEmails(Map<String, String> okapiHeaders,
     Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
-
+    logger.debug("postDelayedTaskRetryFailedEmails::Trying to retry failed emails");
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/DelayedTasksAPI.java:54]: - format: "...:: parameters param1: {} param2: {}"
- headers should be logged

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java [hunk 3]
```diff
@@ -50,11 +51,11 @@ public void deleteDelayedTaskExpiredMessages(String expirationDate, String statu
...
     asyncResultHandler.handle(succeededFuture(
       PostDelayedTaskRetryFailedEmailsResponse.respond202()));
 
-    logger.info("Starting email retry job");
+    logger.info("postDelayedTaskRetryFailedEmails:: Starting email retry job");
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/DelayedTasksAPI.java:58]: I think we can remove this log now that we're logging method name and params above

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java [hunk 4]
```diff
@@ -64,6 +65,7 @@ public void postDelayedTaskRetryFailedEmails(Map<String, String> okapiHeaders,
   }
 
   private Future<List<EmailEntity>> findEmailsForRetry() {
+    logger.debug("findEmailsForRetry::Trying to find emails for retry");
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/DelayedTasksAPI.java:68]: No need to explain what is happening. method name is clear enough

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java [hunk 5]
```diff
@@ -72,15 +74,15 @@ private Future<List<EmailEntity>> findEmailsForRetry() {
...
   }
 
   private static void logRetryResult(AsyncResult<Collection<EmailEntity>> result, long startTimeMillis) {
     long duration = currentTimeMillis() - startTimeMillis;
     if (result.succeeded()) {
-      logger.info("Email retry job took {} ms and finished successfully", duration);
+      logger.info("logRetryResult:: Email retry job took {} ms and finished successfully", duration);
     } else {
-      logger.error("Email retry job took {} ms and failed: {}", duration, result.cause().getMessage());
+      logger.error("logRetryResult:: Email retry job took {} ms and failed: {}", duration, result.cause().getMessage());
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/DelayedTasksAPI.java:85]: result.cause() is Throwable, we should log it with a stacktrace

File: src/main/java/org/folio/rest/impl/EmailAPI.java [hunk 6]
```diff
@@ -24,7 +24,7 @@ public EmailAPI(Vertx vertx, String tenantId) {
   @Override
   public void postEmail(EmailEntity email, Map<String, String> requestHeaders,
     Handler<AsyncResult<Response>> resultHandler, Context vertxContext) {
-
+    logger.debug("postEmail:: parameters: email={}", email);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/EmailAPI.java:27]: How will it log an object such as EmailEntity? What format?

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 7]
```diff
@@ -102,7 +102,7 @@ private void initServices() {
 
   protected Future<EmailEntity> processEmail(EmailEntity email,
     Map<String, String> okapiHeaders) {
-
+    logger.debug("processEmail:: parameter email: {}", email);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:105]: Same question as above

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 8]
```diff
@@ -135,21 +134,21 @@ protected Future<EmailEntity> processEmail(EmailEntity email, SmtpConfiguration
 
   protected EmailEntity handleSuccess(EmailEntity email) {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:135]: Should log method name and parameters

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 9]
```diff
@@ -135,21 +134,21 @@ protected Future<EmailEntity> processEmail(EmailEntity email, SmtpConfiguration
...
     String message = format(SUCCESS_SEND_EMAIL, join(",", email.getTo()));
-    logger.info(message);
+    logger.info("handleSuccess:: {}", message);
 
     return updateEmail(email, DELIVERED, message);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:139]: Should log method result

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 10]
```diff
@@ -135,21 +134,21 @@ protected Future<EmailEntity> processEmail(EmailEntity email, SmtpConfiguration
...
   }
 
   protected EmailEntity handleFailure(EmailEntity email, Throwable throwable) {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:142]: Same here - we should log method name and params when entering the method and the result on exit

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 11]
```diff
@@ -230,30 +229,33 @@ private void deleteEntriesFromModConfig(Configurations configurationsToDelete,
           .send()
           .onSuccess(response -> {
             if (response.statusCode() == HTTP_NO_CONTENT.toInt()) {
-              logger.info("Successfully deleted configuration entry {}", id);
+              logger.info("deleteEntriesFromModConfig:: Successfully deleted configuration entry {}", id);
               return;
             }
-            logger.error(format("Failed to delete configuration entry %s", id));
+            logger.warn(format("deleteEntriesFromModConfig:: Failed to delete configuration entry %s", id));
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:235]: No need to use String.format

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java [hunk 12]
```diff
@@ -51,10 +56,17 @@ public void deleteDelayedTaskExpiredMessages(String expirationDate, String statu
   public void postDelayedTaskRetryFailedEmails(Map<String, String> okapiHeaders,
     Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
 
-    asyncResultHandler.handle(succeededFuture(
-      PostDelayedTaskRetryFailedEmailsResponse.respond202()));
+    logger.debug("postDelayedTaskRetryFailedEmails:: parameters: okapiHeaders={}",
+      () -> logOkapiHeaders(okapiHeaders));
+
+    Handler<AsyncResult<Response>> loggingResponseHandler =
+      loggingResponseHandler("getAutomatedPatronBlocksByUserId", asyncResultHandler, logger);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/DelayedTasksAPI.java:63]: Wrong method name in the log message?

File: src/main/java/org/folio/rest/impl/EmailAPI.java [hunk 13]
```diff
@@ -38,7 +43,7 @@ public void postEmail(EmailEntity email, Map<String, String> requestHeaders,
   public void getEmail(String query, int offset, int limit, String lang,
     Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> resultHandler,
     Context context) {
-
+    logger.debug("getEmail:: parameters: query={}, offset={}, limit={}, lang={}", query, offset, limit, lang);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/EmailAPI.java:46]: Why are we not logging Okapi headers here?

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 14]
```diff
@@ -102,6 +104,8 @@ private void initServices() {
 
   protected Future<EmailEntity> processEmail(EmailEntity email,
     Map<String, String> okapiHeaders) {
+    logger.debug("processEmail:: parameter email: {},requestHeaders={}",
+      () -> logAsJson(email),() -> logOkapiHeaders(okapiHeaders));
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:108]: commas after ,

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 15]
```diff
@@ -102,6 +104,8 @@ private void initServices() {
...
 
     return processEmails(singletonList(email), okapiHeaders)
       .map(emails -> emails.stream().findFirst().orElseThrow());
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:111]: Why are we not logging the result?

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 16]
```diff
@@ -111,19 +115,19 @@ protected Future<Collection<EmailEntity>> processEmails(Collection<EmailEntity>
...
+      logger.info("processEmails:: emails is empty");
       return succeededFuture(emails);
     }
-
-    logger.info("Start processing a batch of {} emails", emails.size());
+    logger.debug("processEmails:: Trying to process a batch of {} emails", emails.size());
 
     return lookupSmtpConfiguration(okapiHeaders)
       .compose(config -> mapInOrder(emails, email -> processEmail(email, config)))
       .recover(t -> handleFailure(emails, t));
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:125]: Not logging the result?

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 17]
```diff
@@ -135,21 +139,25 @@ protected Future<EmailEntity> processEmail(EmailEntity email, SmtpConfiguration
 
   protected EmailEntity handleSuccess(EmailEntity email) {
     String message = format(SUCCESS_SEND_EMAIL, join(",", email.getTo()));
-    logger.info(message);
+    logger.debug("handleSuccess:: parameter email: {}", () -> logAsJson(email));
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:142]: Logging of the method name and parameters should be the first line in the method

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 18]
```diff
@@ -135,21 +139,25 @@ protected Future<EmailEntity> processEmail(EmailEntity email, SmtpConfiguration
...
-    return updateEmail(email, FAILURE, errorMessage);
+    EmailEntity emailEntity = updateEmail(email, FAILURE, errorMessage);
+    logger.info("handleFailure:: result: {}", () -> logAsJson(email));
+    return emailEntity;
   }
 
   private static EmailEntity updateEmail(EmailEntity email, Status status, String message) {
     int newAttemptCount = email.getAttemptCount() + 1;
-
+    logger.debug("updateEmail:: parameters emailId: {}, status: {}, message: {}",() -> logAsJson(email), () -> status, () -> message);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:160]: Comma

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 19]
```diff
@@ -172,6 +180,7 @@ protected Future<Collection<EmailEntity>> handleFailure(Collection<EmailEntity>
   }
 
   private Future<SmtpConfiguration> lookupSmtpConfiguration(Map<String, String> requestHeaders) {
+    logger.debug("lookupSmtpConfiguration:: requestHeaders={}", () -> logOkapiHeaders(requestHeaders));
     return smtpConfigurationService.getSmtpConfiguration()
       .compose(EmailUtils::validateSmtpConfiguration)
       .recover(throwable -> moveConfigsFromModConfigurationToLocalDb(requestHeaders));
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:186]: Not logging the result?

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 20]
```diff
@@ -180,14 +189,15 @@ private Future<SmtpConfiguration> lookupSmtpConfiguration(Map<String, String> re
   private Future<SmtpConfiguration> moveConfigsFromModConfigurationToLocalDb(
     Map<String, String> requestHeaders) {
 
+    logger.debug("moveConfigsFromModConfigurationToLocalDb:: requestHeaders={}", () -> logOkapiHeaders(requestHeaders));
     OkapiClient okapiClient = new OkapiClient(vertx, requestHeaders, webClientOptions);
 
     return fetchSmtpConfigurationFromModConfig(okapiClient)
       .compose(configs -> copyConfigurationAndDeleteFromModConfig(configs, okapiClient));
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:196]: Not logging the result?

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 21]
```diff
@@ -196,12 +206,12 @@ private Future<Configurations> fetchSmtpConfigurationFromModConfig(OkapiClient o
       .compose(response -> {
         if (response.statusCode() == HTTP_OK.toInt()) {
           Configurations config = response.bodyAsJsonObject().mapTo(Configurations.class);
-          logger.info("Successfully fetched {} configuration entries", config.getConfigs().size());
+          logger.info("fetchSmtpConfigurationFromModConfig:: Successfully fetched {} configuration entries", config.getConfigs().size());
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:209]: We can log configs as a list, it's not going to be huge. Also, logList() is going to only log a limited amount of elements

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java (hunk 1 above)
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/impl/DelayedTasksAPI.java]: Still not logging `headers`. Use logOkapiHeaders() for this

File: src/main/java/org/folio/services/email/impl/MailServiceImpl.java [hunk 22]
```diff
@@ -68,17 +70,17 @@ public void sendEmail(JsonObject smtpConfigurationJson, JsonObject emailJson,
...
         .sendMail(mailMessage)
-        .onSuccess(r -> logger.info("Email {} sent in {} ms", emailId, currentTimeMillis() - start))
-        .onFailure(t -> logger.error("Failed to send email {}: {}", emailId, t.getMessage()))
+        .onSuccess(r -> logger.info("sendEmail:: Email {} sent in {} ms", emailId, currentTimeMillis() - start))
+        .onFailure(t -> logger.error("sendEmail:: Failed to send email {}: {}", emailId, t.getMessage()))
         .map(emailJson)
         .onComplete(resultHandler);
     } catch (Exception ex) {
-      logger.error(format(ERROR_SENDING_EMAIL, ex.getMessage()));
+      logger.warn(format(ERROR_SENDING_EMAIL, ex));
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/services/email/impl/MailServiceImpl.java:83]: method name should be added

File: src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java [hunk 23]
```diff
@@ -52,20 +57,25 @@ public void saveEmailEntity(String tenantId, JsonObject emailJson,
...
     }
   }
 
   @Override
   public void findEmailEntries(String tenantId, int limit, int offset, String query,
-                               Handler<AsyncResult<JsonObject>> resultHandler) {
+    Handler<AsyncResult<JsonObject>> resultHandler) {
+
+    logger.debug("findEmailEntries:: parameters: tenantId: {}, limit: {}, offset: {}, query: {}",
+      () -> tenantId, () -> limit, () -> offset, () -> query);
```
Comment (roman-barannyk) [inline src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java:70]: don't have to use suppliers

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 24]
```diff
@@ -263,41 +285,50 @@ protected Future<EmailEntries> findEmailEntries(int limit, int offset, String qu
 
   protected Future<Void> deleteEmailsByExpirationDate(String expirationDate, String emailStatus) {
     Promise<Void> promise = Promise.promise();
+    logger.debug("deleteEmailsByExpirationDate:: parameters expirationDate: {}, emailStatus: {}", expirationDate, emailStatus);
     storageService.deleteEmailEntriesByExpirationDateAndStatus(tenantId, expirationDate, emailStatus,
       result -> {
         if (result.failed()) {
+          logger.warn("deleteEmailsByExpirationDate:: Failed to delete emails with expiration date {}", expirationDate);
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java:292]: should result.cause() be logged as well?

File: src/main/java/org/folio/services/email/impl/MailServiceImpl.java [hunk 25]
```diff
@@ -68,17 +70,17 @@ public void sendEmail(JsonObject smtpConfigurationJson, JsonObject emailJson,
...
         .sendMail(mailMessage)
-        .onSuccess(r -> logger.info("Email {} sent in {} ms", emailId, currentTimeMillis() - start))
-        .onFailure(t -> logger.error("Failed to send email {}: {}", emailId, t.getMessage()))
+        .onSuccess(r -> logger.info("sendEmail:: Email {} sent in {} ms", emailId, currentTimeMillis() - start))
+        .onFailure(t -> logger.error("sendEmail:: Failed to send email {}: {}", emailId, t.getMessage()))
         .map(emailJson)
         .onComplete(resultHandler);
     } catch (Exception ex) {
-      logger.error(format(ERROR_SENDING_EMAIL, ex.getMessage()));
+      logger.warn("sendEmail:: {}", format(ERROR_SENDING_EMAIL, ex));
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/services/email/impl/MailServiceImpl.java:83]: maybe exception with stacktrace should be logged as well?

File: src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java [hunk 26]
```diff
@@ -18,6 +19,7 @@
 import org.folio.rest.persist.interfaces.Results;
 import org.folio.services.storage.StorageService;
 
+
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java:22]: extra blank line

File: src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java [hunk 27]
```diff
@@ -52,20 +57,25 @@ public void saveEmailEntity(String tenantId, JsonObject emailJson,
...
+    logger.debug("findEmailEntries:: parameters tenantId: {}, limit: {}, offset: {}, query: {}",
+      tenantId, limit, offset, query);
     try {
       String[] fieldList = {"*"};
       CQLWrapper cql = getCQL(query, limit, offset);
       PostgresClient pgClient = PostgresClient.getInstance(vertx, tenantId);
       pgClient.get(EMAIL_STATISTICS_TABLE_NAME, EmailEntity.class, fieldList, cql, true, false,
         getReply -> {
           if (getReply.failed()) {
+            logger.warn("findEmailEntries:: Failed to get email entries: {}", getReply::cause);
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java:78]: placeholder not needed

File: src/main/java/org/folio/services/email/impl/MailServiceImpl.java [hunk 28]
```diff
@@ -68,17 +70,17 @@ public void sendEmail(JsonObject smtpConfigurationJson, JsonObject emailJson,
...
-      logger.info("Sending email {}: attempt {}/{}",
+      logger.info("sendEmail:: Sending email {}: attempt {}/{}",
         emailId, emailEntity.getAttemptCount() + 1, RETRY_MAX_ATTEMPTS);
 
       defineMailClient(mailConfig)
         .sendMail(mailMessage)
-        .onSuccess(r -> logger.info("Email {} sent in {} ms", emailId, currentTimeMillis() - start))
-        .onFailure(t -> logger.error("Failed to send email {}: {}", emailId, t.getMessage()))
+        .onSuccess(r -> logger.info("sendEmail:: Email {} sent in {} ms", emailId, currentTimeMillis() - start))
+        .onFailure(t -> logger.error("sendEmail:: Failed to send email {}: {}", emailId, t.getMessage()))
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/services/email/impl/MailServiceImpl.java:79]: log throwable with stacktrace?

File: src/test/java/org/folio/util/LogUtilTest.java [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @Test
+    public void asJsonShouldReturnNullIfCallWithNullAndSizeValue() {
+      assertNull(LogUtil.asJson(null, 10));
+    }
+
+    @Test
+    public void asJsonShouldReturnStringRepresentationOfListOfIntegers() {
+      List<Integer> list = new ArrayList<>();
```
Comment (tatsiana-tarhonskaya) [inline src/test/java/org/folio/util/LogUtilTest.java:46]: List.of?

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java [hunk 30]
```diff
@@ -72,15 +85,15 @@ private Future<List<EmailEntity>> findEmailsForRetry() {
...
   }
 
   private static void logRetryResult(AsyncResult<Collection<EmailEntity>> result, long startTimeMillis) {
     long duration = currentTimeMillis() - startTimeMillis;
     if (result.succeeded()) {
-      logger.info("Email retry job took {} ms and finished successfully", duration);
+      logger.info("logRetryResult:: Email retry job took {} ms and finished successfully", duration);
     } else {
-      logger.error("Email retry job took {} ms and failed: {}", duration, result.cause().getMessage());
+      logger.error("logRetryResult:: Email retry job took {} ms and failed: {}", duration, result.cause());
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/rest/impl/DelayedTasksAPI.java:96]: logger.warn? and there's extra placeholder

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java (hunk 24 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/base/AbstractEmail.java:292]: updated, thanks

File: src/main/java/org/folio/services/email/impl/MailServiceImpl.java (hunk 25 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/services/email/impl/MailServiceImpl.java]: done

File: src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java (hunk 26 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java]: removed

File: src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java (hunk 27 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/services/storage/impl/StorageServiceImpl.java]: done

File: src/main/java/org/folio/services/email/impl/MailServiceImpl.java (hunk 28 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/services/email/impl/MailServiceImpl.java]: done

File: src/test/java/org/folio/util/LogUtilTest.java (hunk 29 above)
Comment (roman-barannyk) [reply src/test/java/org/folio/util/LogUtilTest.java]: updated, thanks

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java (hunk 30 above)
Comment (roman-barannyk) [reply src/main/java/org/folio/rest/impl/DelayedTasksAPI.java]: updated

**Review by alexanderkurash — APPROVED** [review]

**Review by roman-barannyk — APPROVED** [review]

**Pull request merged**